### PR TITLE
Completed CyberSource Payment Processor

### PR DESCRIPTION
**Pull request by clintonb**
This commit includes a complete CyberSource payment processor that aligns with the new order pipeline.

Remaining Task: Include course and track as merchant fields in form parameters

@rlucioni @jimabramson @wedaly

### CODE REVIEW

File: ecommerce/extensions/order/constants.py [hunk 1]
```diff
@@ -0,0 +1,3 @@
+class PaymentEventTypeName(object):
+    PAID = 'Paid'
+    REFUNDED = 'Refunded'
```
Comment (rlucioni) [inline ecommerce/extensions/order/constants.py:3]: Would we benefit from more precise, domain-specific names such as "Settled" and "Credited" which more accurately reflect the transaction types they correspond to? Names like these might give us more flexibility in the future in case handling payment isn't as simple as "Paid."

File: ecommerce/extensions/order/migrations/0005_auto_20150415_0232.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('order', '0004_order_payment_processor'),
+    ]
+
+    operations = [
+        migrations.RemoveField(
+            model_name='order',
+            name='payment_processor',
```
Comment (rlucioni) [inline ecommerce/extensions/order/migrations/0005_auto_20150415_0232.py]: We currently have code in production which relies (naively) on the presence of this field. We should wait to drop it until the code in production no longer relies on it. I've created [XCOM-244](https://openedx.atlassian.net/browse/XCOM-244) to track the work required to deprecate the old version of the ecommerce API; this work falls under it.

File: ecommerce/extensions/order/migrations/0005_auto_20150415_0232.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        ),
+        migrations.AddField(
+            model_name='paymentevent',
+            name='processor_name',
```
Comment (rlucioni) [inline ecommerce/extensions/order/migrations/0005_auto_20150415_0232.py]: Just to make sure I'm on the same page as you: this field is necessary because payment events aren't linked to sources? Sources are linked to source types which include a `name` field for recording processor names, but payment events don't provide any way of recording this same information.

File: ecommerce/extensions/order/constants.py (hunk 1 above)
Comment (clintonb) [inline ecommerce/extensions/order/constants.py:3]: In the future, maybe. For now, we don't have such granularity.

File: ecommerce/extensions/order/models.py [hunk 4]
```diff
@@ -1,13 +1,13 @@
+# noinspection PyUnresolvedReferences
```
Comment (rlucioni) [inline ecommerce/extensions/order/models.py:1]: Boooo! Does this PyCharm silliness need to be here?

File: ecommerce/extensions/order/migrations/0005_auto_20150415_0232.py (hunk 3 above)
Comment (clintonb) [inline ecommerce/extensions/order/migrations/0005_auto_20150415_0232.py]: Correct, there is no linking between the two models. I am still awaiting a response from the mailing list regarding exactly why one would choose to use both `Source` and `PaymentEvent` models.

File: ecommerce/extensions/order/models.py (hunk 4 above)
Comment (clintonb) [inline ecommerce/extensions/order/models.py:1]: Yes. Otherwise, PyCharm's optimization functionality would remove the line.

File: ecommerce/extensions/order/models.py [hunk 5]
```diff
@@ -46,5 +46,9 @@ def check_order_total(cls, order_num, auth_amount, auth_currency):
             return False
 
 
+class PaymentEvent(AbstractPaymentEvent):
+    processor_name = models.CharField(_("Payment Processor"), max_length=32, blank=True, null=True)
```
Comment (rlucioni) [inline ecommerce/extensions/order/models.py:47]: What do you think of naming this field `payment_processor_name`? I tend to avoid leaving "processor" unqualified, in case we ever need to work with another kind of processor. Granted, it is a payment event, so it is clear that this applies to a payment processor.

Comment (clintonb) [inline ecommerce/extensions/order/models.py:47]: That was my initial name; however, it's really long. Given the context of payments, it's pretty safe to assume this name refers to payment processors. If we do manage to need a new processor that is not related to payments (which doesn't seem likely), we can name that field appropriately to avoid confusion.

File: ecommerce/extensions/order/models.py (hunk 4 above)
Comment (rlucioni) [inline ecommerce/extensions/order/models.py:1]: http://www.sublimetext.com/

File: ecommerce/extensions/order/migrations/0005_auto_20150415_0232.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
+# -*- coding: utf-8 -*-
```
Comment (jimabramson) [inline ecommerce/extensions/order/migrations/0005_auto_20150415_0232.py]: any objection to renaming the migration files in this PR to have human-intelligible names?

File: ecommerce/extensions/payment/constants.py [hunk 7]
```diff
@@ -1,49 +1,6 @@
 """Payment processor constants."""
 
+# TODO Use ISO-8601 library?
+ISO_8601_FORMAT = u'%Y-%m-%dT%H:%M:%SZ'
```
Comment (wedaly) [inline ecommerce/extensions/payment/constants.py:3]: Would `datetime.datetime.isoformat()` work?

File: ecommerce/extensions/payment/helpers.py [hunk 8]
```diff
@@ -43,6 +43,6 @@ def sign(message, secret):
     digest = hmac.new(secret, msg=message, digestmod=hashlib.sha256).digest()
 
     # Base64-encode the message hash
-    signature = base64.b64encode(digest).decode()
+    signature = base64.b64encode(digest)
```
Comment (wedaly) [inline ecommerce/extensions/payment/helpers.py]: Why this change?  Does anything downstream assume a unicode object rather than a `str`?

File: ecommerce/extensions/payment/models.py [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+
+class PaymentProcessorResponse(models.Model):
+    """ Auditing model used to save all responses received from payment processors. """
+
+    processor_name = models.CharField(max_length=255, verbose_name=_('Payment Processor'))
+    transaction_id = models.CharField(max_length=255, verbose_name=_('Transaction ID'))
+    basket = models.ForeignKey('basket.Basket', verbose_name=_('Basket'), null=True, blank=True,
+                               on_delete=models.SET_NULL)
+    response = JSONField()
```
Comment (jimabramson) [inline ecommerce/extensions/payment/models.py:13]: are all responses guaranteed to be JSON?  would it be better to make this a generic text field, so that unexpectedly-formatted responses can still be saved in their raw form?

File: ecommerce/extensions/payment/models.py [hunk 10]
```diff
@@ -0,0 +1,21 @@
+from django.db import models
+from django.utils.translation import ugettext_lazy as _
+from jsonfield import JSONField
+
+
+class PaymentProcessorResponse(models.Model):
+    """ Auditing model used to save all responses received from payment processors. """
+
+    processor_name = models.CharField(max_length=255, verbose_name=_('Payment Processor'))
+    transaction_id = models.CharField(max_length=255, verbose_name=_('Transaction ID'))
```
Comment (wedaly) [inline ecommerce/extensions/payment/models.py]: Should we add database indices here?  I suspect we're going to have a lot of lookups by `(processor_name, transaction_id)`

File: ecommerce/extensions/payment/tests/mixins.py [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+        self.assertEqual(order.sources.count(), 1)
+
+        source = order.sources.first()
+        self.assert_basket_matches_source(basket, source, source_type, reference)
+
+
+class CyberSourceMixin(object):
+    """ Mixin with helper methods for testing CyberSource notifications. """
+
+    def generate_signature(self, secret_key, data):
```
Comment (jimabramson) [inline ecommerce/extensions/payment/tests/mixins.py:61]: it seems like `Cybersource._generate_signature` could be adjusted to use this method signature, and then you could use that and avoid having to reimplement the same logic in this mixin.

File: AUTHORS [hunk 12]
```diff
@@ -1,2 +1,4 @@
 Stephen Sanchez <steve@edx.org>
 Renzo Lucioni <renzo@edx.org>
+Clinton Blackburn <cblackburn@edx.org>
+Jim Abramson <jsa@edx.org>
```
Comment (jimabramson) [inline AUTHORS:4]: @clintonb thanks :-)

File: ecommerce/extensions/payment/views.py [hunk 13]
```diff
@@ -1,61 +1,143 @@
...
-        order.set_status(ORDER.PAID)
+        # Note (CCB): No data should be persisted to the database until the payment processor has validated the
+        # response's signature. This validation is performed in the handle_payment method. After that method succeeds,
+        # the response can be safely assumed to have originated from CyberSource.
+        cybersource_response = request.POST.dict()
+
+        # Don't do anything if the response is not valid.
+        if not self.payment_processor.is_response_valid(cybersource_response):
+            logger.info(u'Received invalid CyberSource response: %s', json.dumps(cybersource_response))
+            return HttpResponse()
```
Comment (jimabramson) [inline ecommerce/extensions/payment/views.py]: I think we should respond with status 400 here.

Comment (jimabramson) [inline ecommerce/extensions/payment/views.py]: I also think we should consider recording the response somewhere more durable (i.e. database) so it is accessible/replayable even if it didn't pass validation.  This would avoid losing rejected transactions data in the event of a misconfigured key or other intermittent code/config bug.  Maybe using the payment event table with a processor name of 'invalid' would make sense for this?

**Comment by jimabramson** [comment]
@clintonb bonus points for (so far) removing more lines than you've added :-)

File: ecommerce/extensions/payment/constants.py (hunk 7 above)
Comment (rlucioni) [inline ecommerce/extensions/payment/constants.py:3]: According to the data type definitions presented in the Secure Acceptance docs, CyberSource expects a datetime string formatted exactly this way. `datetime.datetime.isoformat()` includes microseconds and doesn't append the zone designator for the zero UTC offset (`Z`).

File: ecommerce/extensions/order/migrations/0005_auto_20150415_0232.py (hunk 2 above)
Comment (clintonb) [inline ecommerce/extensions/order/migrations/0005_auto_20150415_0232.py]: Removed.

File: ecommerce/extensions/order/migrations/0005_auto_20150415_0232.py (hunk 6 above)
Comment (clintonb) [inline ecommerce/extensions/order/migrations/0005_auto_20150415_0232.py]: Done. To avoid long names, I have split these two operations into separate migrations.

File: ecommerce/extensions/order/migrations/0005_auto_20150415_0232.py (hunk 2 above)
Comment (clintonb) [inline ecommerce/extensions/order/migrations/0005_auto_20150415_0232.py]: Correction: the field is now nullable.

File: ecommerce/extensions/payment/constants.py (hunk 7 above)
Comment (clintonb) [inline ecommerce/extensions/payment/constants.py:3]: Correct. CyberSource doesn't fully support ISO8601, which allows for a timezone offset (`+00:00`). The `signed_date_time` field is explicitly limited to 20 characters.

File: ecommerce/extensions/payment/processors.py [hunk 14]
```diff
@@ -1,44 +1,76 @@
...
+    @abc.abstractmethod
+    def get_transaction_parameters(self, basket, receipt_page_url=None, cancel_page_url=None, **kwargs):
+        """
+        Generate a dictionary of signed parameters required for this processor to complete a transaction.
+
+        Arguments:
+            basket (Basket): The basket of products being purchased..
+
+        Keyword Arguments:
+            receipt_page_url (unicode): If provided, overrides the receipt page URL on the Secure Acceptance
```
Comment (rlucioni) [inline ecommerce/extensions/payment/processors.py]: I don't think `BasePaymentProcessor` should make any mention of or concessions to Secure Acceptance, or to any specific payment processor for that matter. Could you describe `receipt_page_url` and `cancel_page_url` more generally? It looks like [PayPal](https://developer.paypal.com/docs/api/#redirecturls-object) will also be able to make use of them, so we should try to avoid suggesting that these arguments are CyberSource-specific.

File: ecommerce/extensions/payment/processors.py [hunk 15]
```diff
@@ -74,18 +123,14 @@ def __init__(self):
...
-            cancel_page_url=None,
-            merchant_defined_data=None
-    ):
-        """Generate a dictionary of signed parameters CyberSource requires to complete a transaction.
+    def get_transaction_parameters(self, basket, receipt_page_url=None, cancel_page_url=None, **kwargs):
+        """
+        Generate a dictionary of signed parameters CyberSource requires to complete a transaction.
 
         Arguments:
             order (Order): The order whose line items are to be purchased as part of the transaction.
```
Comment (rlucioni) [inline ecommerce/extensions/payment/processors.py]: Could you update this part of the docstring? This method takes baskets now, not orders.

File: ecommerce/extensions/payment/helpers.py (hunk 8 above)
Comment (clintonb) [inline ecommerce/extensions/payment/helpers.py]: Restored. I was having issues last night POSTing to CyberSource. This removal was one of my attempted solutions.

File: ecommerce/extensions/payment/models.py (hunk 10 above)
Comment (clintonb) [inline ecommerce/extensions/payment/models.py]: I'm inclined to say yes. However, there are currently no use cases for reading this data within Oscar.

File: ecommerce/extensions/payment/models.py (hunk 9 above)
Comment (clintonb) [inline ecommerce/extensions/payment/models.py:13]: All data for CyberSource and PayPal will be either JSON or form data. Regardless of how we receive the data, it is all converted to a `dict` for consumption by the payment processors. That `dict` is converted to JSON to aid any future investigations that may need to parse the data.

File: ecommerce/extensions/payment/processors.py [hunk 16]
```diff
@@ -95,176 +140,109 @@ def get_transaction_parameters(
...
+
+        source = Source(source_type=source_type,
+                        currency=currency,
+                        amount_allocated=total,
+                        amount_debited=total,
+                        reference=transaction_id)
+
+        # Create PaymentEvent to track
+        event_type, __ = PaymentEventType.objects.get_or_create(name=PaymentEventTypeName.PAID)
+        event = PaymentEvent(event_type=event_type, amount=total, reference=transaction_id, processor_name=self.NAME)
```
Comment (rlucioni) [inline ecommerce/extensions/payment/processors.py:239]: From what I've seen in the Oscar docs, these operations (creation of a Source and PaymentEvent) are typically performed with the `handle_payment` method from the Checkout app's `OrderPlacementMixin`. Here's [an example](https://github.com/edx/ecommerce/blob/renzo/handle-payment-processor-response/ecommerce/extensions/checkout/mixins.py#L18) from an older branch of mine. Any reason you're not doing that here?

Comment (rlucioni) [inline ecommerce/extensions/payment/processors.py:239]: Using the `add_payment_source` and `add_payment_event` methods provided by the `OrderPlacementMixin` also pairs nicely with use of the `handle_order_placement` method, also provided by the `OrderPlacementMixin`. Sources and payment events must be linked to an order, and the `handle_order_placement` method takes care of attaching an order to these models before saving them.

File: ecommerce/extensions/payment/models.py (hunk 10 above)
Comment (clintonb) [inline ecommerce/extensions/payment/models.py]: Compound index added. As we process more transactions, it will be easier to remove the unnecessary index than it will be to add the index for an ever-expanding table.

File: ecommerce/extensions/payment/tests/mixins.py (hunk 11 above)
Comment (clintonb) [inline ecommerce/extensions/payment/tests/mixins.py:61]: Hmm...yes and no. I am leary of sharing too much code between the test and the implementation being tested. Happy to discuss this if you have thoughts. I'm already breaking this rule by sharing the `helpers.sign` method.

File: ecommerce/extensions/payment/views.py (hunk 13 above)
Comment (clintonb) [inline ecommerce/extensions/payment/views.py]: I thought about returning a different status. The CyberSource docs make no mention of expected statuses. Given that all of this code is public my argument for not providing information to potential attackers is moot. My only concern regarding recording the information is the potential to have spam in the database, since everything posted would be stored. Thoughts on this?

File: ecommerce/extensions/order/models.py (hunk 4 above)
Comment (clintonb) [inline ecommerce/extensions/order/models.py:1]: https://www.jetbrains.com/pycharm/

File: ecommerce/extensions/payment/processors.py (hunk 14 above)
Comment (clintonb) [inline ecommerce/extensions/payment/processors.py]: Agreed. Updated.

File: ecommerce/extensions/payment/processors.py (hunk 15 above)
Comment (clintonb) [inline ecommerce/extensions/payment/processors.py]: Updated.

File: ecommerce/extensions/payment/processors.py (hunk 16 above)
Comment (clintonb) [inline ecommerce/extensions/payment/processors.py:239]: This is a GitHub bug. The source and payment events are created in `handle_processor_response` which is called by `handle_payment`.

File: ecommerce/extensions/payment/views.py [hunk 17]
```diff
@@ -1,61 +1,143 @@
...
-        # It doesn't matter how we respond to the payment processor if the
-        # payment failed.
-        return HttpResponse()
+    @method_decorator(csrf_exempt)
+    def dispatch(self, request, *args, **kwargs):
+        self.payment_processor = processors.Cybersource()
+        return super(CyberSourceNotifyView, self).dispatch(request, *args, **kwargs)
 
-    def _register_payment(self, order, processor_name):
+    def add_payment_event(self, event):
```
Comment (rlucioni) [inline ecommerce/extensions/payment/views.py]: What's your reasoning behind tearing the source and payment event management logic out of the `OrderPlacementMixin`? You've split the work done by the mixin across this view and the CyberSource processor class, but it's not clear to me why that's necessary.

File: ecommerce/extensions/payment/views.py [hunk 18]
```diff
@@ -1,61 +1,143 @@
...
-            self._register_payment(order, payment_processor.NAME)
-            # fulfill the order
-            self._fulfill_order(order)
-
-        # It doesn't matter how we respond to the payment processor if the
-        # payment failed.
-        return HttpResponse()
+    @method_decorator(csrf_exempt)
+    def dispatch(self, request, *args, **kwargs):
+        self.payment_processor = processors.Cybersource()
```
Comment (rlucioni) [inline ecommerce/extensions/payment/views.py]: Why is it necessary to rely on `dispatch()` to set `payment_processor`?

File: ecommerce/extensions/payment/views.py [hunk 19]
```diff
@@ -1,61 +1,143 @@
...
+
+        # If we don't have a basket, we cannot continue.
+        if not basket:
+            return HttpResponse()
+
+        order_number = self.generate_order_number(basket)
+        currency = cybersource_response['req_currency']
+        _tax = Decimal(cybersource_response.get('req_tax_amount', 0))
+        _total = Decimal(cybersource_response['req_amount'])
+        order_total = prices.Price(currency=currency, excl_tax=_total - _tax, incl_tax=_total)
```
Comment (rlucioni) [inline ecommerce/extensions/payment/views.py]: I'd suggest using Oscar's `OrderTotalCalculator` to compute the order total. Here's [an example](https://github.com/edx/ecommerce/commit/67c120efdc35572f79e5e1121f84888a40eb9dd3#diff-2b9dcec2f6e9929122bcf5071e564312R78). It takes the basket and chosen shipping method into account when calculating the order total.

File: ecommerce/extensions/payment/views.py [hunk 20]
```diff
@@ -1,61 +1,143 @@
...
+                                         # Oscar uses line4 for city
+                                         line4=cybersource_response['req_bill_to_address_city'],
+                                         postcode=cybersource_response['req_bill_to_address_postal_code'],
+                                         state=cybersource_response['req_bill_to_address_state'],
+                                         country=Country.objects.get(
+                                             iso_3166_1_a2=cybersource_response['req_bill_to_address_country']))
+
+        try:
+            user = basket.owner
+            self.handle_order_placement(order_number, user, basket, None, shipping_method, shipping_charge,
```
Comment (rlucioni) [inline ecommerce/extensions/payment/views.py:115]: The `OrderPlacementMixin` provides a method `get_initial_order_status` which is called to populate an order's initial status upon creation. It does nothing be default; I'd suggest overriding the method to return the `OPEN` status. If you don't intend to override that method (or the status it provides isn't appropriate), you may want to pass the `status` kwarg when calling `handle_order_placement`.

File: ecommerce/extensions/payment/views.py (hunk 18 above)
Comment (clintonb) [inline ecommerce/extensions/payment/views.py]: To ensure that it is always set.

Comment (rlucioni) [inline ecommerce/extensions/payment/views.py]: Since setting it doesn't require parsing the request, wouldn't it always be set when the class is instantiated if you were just to set it as a class variable on line 34?

File: ecommerce/extensions/payment/views.py (hunk 17 above)
Comment (clintonb) [inline ecommerce/extensions/payment/views.py]: My intention is to move more of payment-related logic to the processor itself, and keep the view pretty dumb. Give me a moment to think about this some more.

File: ecommerce/extensions/payment/views.py (hunk 19 above)
Comment (clintonb) [inline ecommerce/extensions/payment/views.py]: In this case we explicitly want the amounts authorized by the payment processor; although, we aren't actually using them.

Comment (rlucioni) [inline ecommerce/extensions/payment/views.py]: Got it, thanks for the clarification.

File: ecommerce/extensions/payment/views.py (hunk 20 above)
Comment (clintonb) [inline ecommerce/extensions/payment/views.py:115]: Updated.

Comment (clintonb) [inline ecommerce/extensions/payment/views.py:115]: Actually, the underlying code uses `settings.OSCAR_INITIAL_ORDER_STATUS` to retrieve the state. I'm going to rely on this to avoid any future confusion. That being said, I have updated the tests to ensure the order's status is `'OPEN'`.

File: ecommerce/extensions/payment/views.py (hunk 19 above)
Comment (clintonb) [inline ecommerce/extensions/payment/views.py]: Thought about this further and implemented your suggestion. Additionally, I have updated `handle_payment` to only require the `response` and `basket` parameters. Thanks!

File: ecommerce/extensions/payment/tests/mixins.py (hunk 11 above)
Comment (jimabramson) [inline ecommerce/extensions/payment/tests/mixins.py:61]: in general i think the concern would be addressed by ensuring tests exist for the method in question - if the test validates that it works correctly, then you could feel free to use that method directly elsewhere in your test code.

File: ecommerce/extensions/payment/views.py (hunk 13 above)
Comment (jimabramson) [inline ecommerce/extensions/payment/views.py]: In the case of http status, i was thinking about that from an operational point of view rather than from the client's.  If there was a burst of invalid callback requests in prod, the way things stand we would just have happy 200's at the web server, and would only notice problems if we happened to be reading lines logged at INFO (not likely!)   If the server treats these as errors we're more likely to notice them, and more quickly.

Comment (jimabramson) [inline ecommerce/extensions/payment/views.py]: regarding saving responses to the database, yes, agreed there's a risk we're just gathering spam.  not sure what the industry best practice is here, but the scenario i'm imagining is where something gets misconfigured due to an honest mistake, and we start losing important messages about payments.  the recovery would be either reconciling directly with the payment processor, for which i assume there is some offline workflow (?), or re-processing the missed messages if we have them in our possession.

File: ecommerce/extensions/payment/views.py (hunk 18 above)
Comment (clintonb) [inline ecommerce/extensions/payment/views.py]: I had some initial concerns about class loading issues, but those are unfounded. Updated.

File: ecommerce/extensions/payment/views.py (hunk 13 above)
Comment (clintonb) [inline ecommerce/extensions/payment/views.py]: I have made a few changes:
- All responses are saved to the database...unless the keys are completely wrong (e.g. no signature, transaction_id key)
- If the keys are completely wrong (e.g. spam), we return a 500
- If the signature is invalid we return a 400, and log an ERROR.
- All other combinations return 200.
- All responses that are not representative of an accepted payment (e.g. declined) are logged at the ERROR level.

Using this scheme we are recording all responses that can be reasonably assumed to have come from CyberSource, and logging at an appropriate severity level to Splunk.

File: ecommerce/extensions/payment/views.py (hunk 20 above)
Comment (rlucioni) [inline ecommerce/extensions/payment/views.py:115]: It's worth noting that without redefining the order status pipeline, only orders in the `PAID` state are eligible for fulfillment. I've gone through the exercise of redefining the order status pipeline [here](https://github.com/edx/ecommerce/compare/renzo/handle-payment-processor-response#diff-450d651868a8f1ab20c1dad5c5582c68R39). It might be appropriate to include this redefinition in your PR; otherwise, you can use the `PAID` state as the initial order status, and we can introduce this change in a separate PR.

File: ecommerce/extensions/fulfillment/apps.py [hunk 21]
```diff
@@ -0,0 +1,8 @@
+from django.apps import AppConfig
+
+
+class FulffillmentAppConfig(AppConfig):
```
Comment (rlucioni) [inline ecommerce/extensions/fulfillment/apps.py]: `s/FulffillmentAppConfig/FulfillmentAppConfig`

File: ecommerce/extensions/payment/models.py [hunk 22]
```diff
@@ -0,0 +1,24 @@
...
+    basket = models.ForeignKey('basket.Basket', verbose_name=_('Basket'), null=True, blank=True,
+                               on_delete=models.SET_NULL)
+    response = JSONField()
+    created = models.DateTimeField(auto_now_add=True, db_index=True)
+
+    class Meta(object):
+        index_together = (
+            ('processor_name', 'transaction_id')
+        )
```
Comment (rlucioni) [inline ecommerce/extensions/payment/models.py]: Minor: As of Django 1.7, `index_together` can be a single list when dealing with a single set of fields ([reference](https://docs.djangoproject.com/en/1.7/ref/models/options/#index-together)):

``` python
index_together = ('processor_name', 'transaction_id')
```

File: ecommerce/extensions/fulfillment/apps.py (hunk 21 above)
Comment (clintonb) [inline ecommerce/extensions/fulfillment/apps.py]: Corrected.

File: ecommerce/extensions/payment/models.py (hunk 22 above)
Comment (clintonb) [inline ecommerce/extensions/payment/models.py]: I am aware of this.

File: ecommerce/extensions/payment/processors.py [hunk 23]
```diff
@@ -75,197 +121,123 @@ def __init__(self):
...
+            u'signed_date_time': datetime.datetime.utcnow().strftime(ISO_8601_FORMAT),
+            u'locale': self.language_code,
+            u'transaction_type': u'sale',
+            u'reference_number': unicode(basket.id),
+            u'amount': unicode(basket.total_incl_tax),
+            u'currency': basket.currency,
+            u'consumer_id': basket.owner.username
+        }
+
+        # TODO Include edX-specific data (e.g. course_id, seat type)
```
Comment (rlucioni) [inline ecommerce/extensions/payment/processors.py:159]: Doing this for CyberSource would seem to require the ability to include "merchant-defined data" in the payment parameters, but you've removed that code. Would you be opposed to retaining the ability to add "merchant-defined data" to the payment parameters, so that we can easily add extra information in the future?

Comment (clintonb) [inline ecommerce/extensions/payment/processors.py:159]: I'm not at all opposed to this. It's on my list of things to do.

File: ecommerce/extensions/payment/views.py [hunk 24]
```diff
@@ -1,61 +1,140 @@
...
+        """ Handle the response we've been given from the processor. """
 
-        # update the status of the order
-        order.set_status(ORDER.PAID)
+        # Note (CCB): Orders should not be created until the payment processor has validated the response's signature.
+        # This validation is performed in the handle_payment method. After that method succeeds, the response can be
+        # safely assumed to have originated from CyberSource.
+        cybersource_response = request.POST.dict()
+
+        basket_id = cybersource_response['req_reference_number']
```
Comment (rlucioni) [inline ecommerce/extensions/payment/views.py]: Requests passed to this view without a `req_reference_number` parameter in the body will trigger a `KeyError` here. I'd suggest using `get()`, then either catching `TypeError` with the exception handler (raised when you attempt to convert `None` to an `int`) or explicitly checking if `basket_id` is `None` before proceeding.

File: ecommerce/extensions/payment/views.py [hunk 25]
```diff
@@ -1,61 +1,140 @@
...
+            basket = Basket.objects.get(id=basket_id)
+            basket.strategy = strategy.Default()
+        except (ValueError, ObjectDoesNotExist):
+            logger.exception('Received payment for non-existent basket [%s].', basket_id)
+
+            # Set the basket to None to trigger the return logic below, after we record the processor response.
+            basket = None
+
+        # Store the response in the database
+        transaction_id = cybersource_response['transaction_id']
```
Comment (rlucioni) [inline ecommerce/extensions/payment/views.py]: For the reasons cited above, I'd also use `get()` here and in every other place below where you need to retrieve a value from the response.

File: ecommerce/extensions/payment/processors.py (hunk 23 above)
Comment (rlucioni) [inline ecommerce/extensions/payment/processors.py:159]: My bad, I forgot that it was on the list you posted at the outset.

File: ecommerce/extensions/payment/views.py [hunk 26]
```diff
@@ -1,61 +1,143 @@
...
+                basket = Basket.objects.get(id=basket_id)
+                basket.strategy = strategy.Default()
+            except (ValueError, ObjectDoesNotExist):
+                logger.exception('Received payment for non-existent basket [%s].', basket_id)
+
+                # Set the basket to None to trigger the return logic below, after we record the processor response.
+                basket = None
+        finally:
+            # Store the response in the database
+            ppr = self.payment_processor.record_processor_response(transaction_id, cybersource_response, basket=basket)
```
Comment (rlucioni) [inline ecommerce/extensions/payment/views.py]: Since the `transaction_id` field on PaymentProcessorResponse is not nullable, this will raise an `IntegrityError` if `transaction_id` continues to be `None` at this point. Maybe you could check that `transaction_id` has a value other than `None` before recording the processor response? This would also serve to keep some spam out of our database, since any response from CyberSource should contain a `transaction_id`.

File: ecommerce/extensions/payment/processors.py [hunk 27]
```diff
@@ -75,197 +121,123 @@ def __init__(self):
...
+        source = Source(source_type=source_type,
+                        currency=currency,
+                        amount_allocated=total,
+                        amount_debited=total,
+                        reference=transaction_id,
+                        label=req_card_number)
+
+        # Create PaymentEvent to track
+        event_type, __ = PaymentEventType.objects.get_or_create(name=PaymentEventTypeName.PAID)
+        event = PaymentEvent(event_type=event_type, amount=total, reference=transaction_id, processor_name=self.NAME)
```
Comment (rlucioni) [inline ecommerce/extensions/payment/processors.py:239]: It looks like the source and payment event creation logic here would apply equally well to PayPal and other processors. What's your opinion on overriding `add_payment_source` and `add_payment_event` on the `EdxOrderPlacementMixin` itself to have them do the work you're doing here?

File: ecommerce/extensions/payment/views.py [hunk 28]
```diff
@@ -1,61 +1,143 @@
...
 
-        # record payment events
         self.add_payment_source(source)
-        self.add_payment_event(PC.PAID_EVENT_NAME, order.total_excl_tax, order.number)
-        self.save_payment_details(order)
+        self.add_payment_event(payment_event)
+
+    def handle_successful_order(self, order):
+        # Send a signal so that receivers can perform relevant tasks (e.g. fulfill the order).
+        post_checkout.send(sender=self, order=order)
```
Comment (rlucioni) [inline ecommerce/extensions/payment/views.py]: This also looks like logic that we'd benefit from sharing with other processor workflows. You could override `handle_successful_order` on the `EdxOrderPlacementMixin` instead of doing it here.

File: ecommerce/extensions/payment/views.py [hunk 29]
```diff
@@ -1,61 +1,143 @@
...
-        source_type, _ = SourceType.objects.get_or_create(name=processor_name)
-        source = source_type.sources.model(
-            source_type=source_type, amount_allocated=order.total_excl_tax, currency=order.currency
-        )
+        This method is responsible for handling payment and recording the
+        payment sources (using the add_payment_source method) and payment
+        events (using add_payment_event) so they can be
+        linked to the order when it is saved later on.
+        """
+        source, payment_event = self.payment_processor.handle_processor_response(response, basket=basket)
```
Comment (rlucioni) [inline ecommerce/extensions/payment/views.py]: I'll lay off after this one, but this also looks like a method that we'd benefit from overriding on the `EdxOrderPlacementMixin`. You could redefine the signature slightly to accept a `payment_processor`, or just take advantage of the `**kwargs` already included in the signature.

File: ecommerce/extensions/payment/views.py (hunk 26 above)
Comment (clintonb) [inline ecommerce/extensions/payment/views.py]: `transaction_id` is now nullable. We still want to record spam responses in case we need to perform an investigation. Over time they can be easily pruned by removing those with a null transaction_id.

File: ecommerce/extensions/payment/views.py (hunk 28 above)
Comment (clintonb) [inline ecommerce/extensions/payment/views.py]: Added.

File: ecommerce/extensions/payment/views.py (hunk 29 above)
Comment (clintonb) [inline ecommerce/extensions/payment/views.py]: I've removed the `handle_payment` override in `EdxOrderPlacementMixin` since it wasn't actually being used.

File: ecommerce/extensions/payment/processors.py (hunk 27 above)
Comment (clintonb) [inline ecommerce/extensions/payment/processors.py:239]: Given the specificity of the `Source` as it relates to the payment processor, I have opted to keep the creation of both `PaymentEvent` and `Source` in this class. We can revisit this implementation when we start work on PayPal.

Comment (rlucioni) [inline ecommerce/extensions/payment/processors.py:239]: Sounds good to me.

File: ecommerce/extensions/payment/views.py [hunk 30]
```diff
@@ -0,0 +1,122 @@
...
+            # Store the response in the database regardless of its authenticity.
+            ppr = self.payment_processor.record_processor_response(cybersource_response, transaction_id=transaction_id,
+                                                                   basket=basket)
+
+        try:
+            self.handle_payment(cybersource_response, basket)
+        except InvalidSignatureError:
+            logger.exception(
+                'Received an invalid CyberSource response. The payment response was recorded in entry [%d].', ppr.id)
+            return HttpResponse(status=400)
```
Comment (rlucioni) [inline ecommerce/extensions/payment/views.py:94]: Minor: For consistency with our other views, I'd suggest using DRF's `HTTP_400_BAD_REQUEST` status constant. That said, this isn't a DRF view, so this is fine, too.

File: ecommerce/extensions/payment/tests/mixins.py [hunk 31]
```diff
@@ -0,0 +1,103 @@
...
+    def assert_payment_source_exists(self, basket, source_type, reference, label):
+        """ Validates that a single Source exists for the basket's associated order. """
+        order = self.get_order(basket)
+        self.assertEqual(order.sources.count(), 1)
+
+        source = order.sources.first()
+        self.assert_basket_matches_source(basket, source, source_type, reference, label)
+
+
+class CyberSourceMixin(object):
```
Comment (rlucioni) [inline ecommerce/extensions/payment/tests/mixins.py]: Minor: Our other CyberSource class names lowercase the "s": `CybersourceMixin`.

**Comment by rlucioni** [comment]
This looks great! Thanks for taking it on. :+1:
